num/dualcmplx/dual.go

```diff
@@ -0,0 +1,269 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package dualcmplx
+
+import (
+	"fmt"
+	"math"
+	"math/cmplx"
+	"strings"
+
+	"gonum.org/v1/gonum/num/dual"
+)
+
+// Number is a float64 precision anti-commutative dual complex number.
+type Number struct {
+	Real, Dual complex128
+}
+
+var zero Number
+
+// Format implements fmt.Formatter.
+func (d Number) Format(fs fmt.State, c rune) {
+	prec, pOk := fs.Precision()
+	if !pOk {
+		prec = -1
+	}
+	width, wOk := fs.Width()
+	if !wOk {
+		width = -1
+	}
+	switch c {
+	case 'v':
+		if fs.Flag('#') {
+			fmt.Fprintf(fs, "%T{Real:%#v, Dual:%#v}", d, d.Real, d.Dual)
+			return
+		}
+		if fs.Flag('+') {
+			fmt.Fprintf(fs, "{Real:%+v, Dual:%+v}", d.Real, d.Dual)
+			return
+		}
+		c = 'g'
+		prec = -1
+		fallthrough
+	case 'e', 'E', 'f', 'F', 'g', 'G':
+		fre := fmtString(fs, c, prec, width, false)
+		fim := fmtString(fs, c, prec, width, true)
+		fmt.Fprintf(fs, fmt.Sprintf("(%s+%[2]sϵ)", fre, fim), d.Real, d.Dual)
+	default:
+		fmt.Fprintf(fs, "%%!%c(%T=%[2]v)", c, d)
+		return
+	}
+}
+
+// This is horrible, but it's what we have.
+func fmtString(fs fmt.State, c rune, prec, width int, wantPlus bool) string {
+	var b strings.Builder
+	b.WriteByte('%')
+	for _, f := range "0+- " {
+		if fs.Flag(int(f)) || (f == '+' && wantPlus) {
+			b.WriteByte(byte(f))
+		}
+	}
+	if width >= 0 {
+		fmt.Fprint(&b, width)
+	}
+	if prec >= 0 {
+		b.WriteByte('.')
+		if prec > 0 {
+			fmt.Fprint(&b, prec)
+		}
+	}
+	b.WriteRune(c)
+	return b.String()
+}
+
+// Add returns the sum of x and y.
+func Add(x, y Number) Number {
+	return Number{
+		Real: x.Real + y.Real,
+		Dual: x.Dual + y.Dual,
+	}
+}
+
+// Sub returns the difference of x and y, x-y.
+func Sub(x, y Number) Number {
+	return Number{
+		Real: x.Real + y.Real,
+		Dual: x.Dual + y.Dual,
+	}
+}
+
+// Mul returns the dual product of x and y.
```

Comment:
, x*y (order or multiplication is important so be explicit)

Reply:
Done.